neps/optimizers/models/deepGP.py

```diff
@@ -0,0 +1,572 @@
+            ),
+        )
+        self.cnn = nn.Sequential(
+            nn.Conv1d(
+                in_channels=1,
+                kernel_size=(kwargs.get("cnn_kernel_size", 3),),
+                out_channels=4,
+            ),
+            nn.AdaptiveMaxPool1d(1),
+        )
+
+    def forward(self, x, budgets, learning_curves):
+        # add an extra dimensionality for the budget
+        # making it nr_rows x 1.
+        budgets = torch.unsqueeze(budgets, dim=1)
+        # concatenate budgets with examples
+        x = torch.cat((x, budgets), dim=1)
+        x = self.fc1(x)
+        x = self.activation(self.bn1(x))
+
+        for i in range(2, self.n_layers):
+            x = self.activation(getattr(self, f"bn{i}")(getattr(self, f"fc{i}")(x)))
+
+        # add an extra dimensionality for the learning curve
+        # making it nr_rows x 1 x lc_values.
+        learning_curves = torch.unsqueeze(learning_curves, 1)
+        lc_features = self.cnn(learning_curves)
+        # revert the output from the cnn into nr_rows x nr_kernels.
+        lc_features = torch.squeeze(lc_features, 2)
+
+        # put learning curve features into the last layer along with the higher level features.
+        x = torch.cat((x, lc_features), dim=1)
+        x = self.activation(getattr(self, f"fc{self.n_layers}")(x))
+
+        return x
+
+
+class GPRegressionModel(gpytorch.models.ExactGP):
+    """
+    A simple GP model.
+    """
+
+    def __init__(
+        self,
+        train_x: torch.Tensor,
+        train_y: torch.Tensor,
+        likelihood: gpytorch.likelihoods.GaussianLikelihood,
+    ):
+        """
+        Constructor of the GPRegressionModel.
+
+        Args:
+            train_x: The initial train examples for the GP.
+            train_y: The initial train labels for the GP.
+            likelihood: The likelihood to be used.
+        """
+        super().__init__(train_x, train_y, likelihood)
+
+        self.mean_module = gpytorch.means.ConstantMean()
+        self.covar_module = gpytorch.kernels.ScaleKernel(gpytorch.kernels.RBFKernel())
+
+    def forward(self, x):
+        mean_x = self.mean_module(x)
+        covar_x = self.covar_module(x)
+
+        return gpytorch.distributions.MultivariateNormal(mean_x, covar_x)
+
+
+@dataclass
+class DeepGPDataTransformer:
+    # TODO: This class could be used for other models as well
+    space: SearchSpace
+    fidelity_bounds: tuple[int | float, int | float] | None
+    normalize_y: bool
+    min_learning_curve_length: int
+    learning_curve_pad_value: float
+    device: torch.device
+
+    numericals: dict[str, FloatParameter | IntegerParameter] = field(init=False)
+    categoricals: dict[str, CategoricalParameter] = field(init=False)
+    output_dim: int = field(init=False)
+
+    def __post_init__(self) -> None:
+        self.numericals = {
+            name: h
+            for name, h in self.space.items()
+            if isinstance(h, (FloatParameter, IntegerParameter))
```

Comment:
The current implementation calls for separating the fidelity parameter from the rest of the parameters but if we change the `NeuralFeatureExtractor.forward(...)` above, and normalize fidelity correctly, we can keep fidelity parameter together with the other numericals

Reply:
Oh good spot, this should have been ignoring fidelities in there

Reply:
The refactored NN i will leave out until we can verify its performance in some way